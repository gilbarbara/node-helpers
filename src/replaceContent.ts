/* eslint-disable no-await-in-loop */
import { readFile, writeFile } from 'node:fs/promises';
import { relative, resolve } from 'node:path';

import fg from 'fast-glob';

export type ReplaceActionResult = string | false | undefined;

export interface ReplaceContentOptions {
  callback(content: string): Promise<ReplaceActionResult> | ReplaceActionResult;
  directory?: string;
  globOptions?: fg.Options;
  name?: string;
  pattern: fg.Pattern | fg.Pattern[];
  silent?: boolean;
}

export async function replaceContent(options: ReplaceContentOptions) {
  const {
    callback,
    directory = 'dist',
    globOptions = {},
    name = 'replaceContent',
    pattern,
    silent,
  } = options;
  const cwd = process.cwd();

  const files = await fg(pattern, {
    cwd: resolve(cwd, directory),
    ...globOptions,
    absolute: true,
    ignore: [...(globOptions.ignore || []), 'node_modules'],
  });

  const logger = (message: string) => {
    if (!silent) {
      // eslint-disable-next-line no-console
      console.log(`[${name}] ${message}`);
    }
  };

  if (!files.length) {
    logger('No files matched');
  }

  const output = [];

  for (const file of files) {
    const content = await readFile(file, 'utf8');
    const result = await callback(content);
    const filename = relative(cwd, file);

    if (result) {
      await writeFile(file, result);
      output.push(result);
      logger(`âœ“ ${filename}`);
    } else {
      logger(`skip ${filename}`);
    }
  }

  return output;
}
