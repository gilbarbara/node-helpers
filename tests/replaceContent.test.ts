import { copyFile, mkdir, unlink } from 'node:fs/promises';
import { resolve } from 'node:path';

import { SpyInstance, vi } from 'vitest';

import { replaceContent } from '../src/replaceContent';

const tempDirectory = resolve(__dirname, '.tmp');

describe('replaceContent', () => {
  let consoleLog: SpyInstance;

  beforeAll(async () => {
    await mkdir(tempDirectory, { recursive: true });
    await copyFile(
      resolve(__dirname, '__fixtures__/replaceContent.d.ts'),
      `${tempDirectory}/replaceContent.d.ts`,
    );

    consoleLog = vi.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    consoleLog.mockClear();
  });

  afterAll(async () => {
    await unlink(`${tempDirectory}/replaceContent.d.ts`);
    consoleLog.mockRestore();
  });

  it('should replace properly', async () => {
    const result = await replaceContent({
      directory: 'tests/.tmp',
      name: 'fix-cjs-dts',
      pattern: '**/*.d.ts',
      callback: content => {
        const { imp } = /(?<imp>import .+ from 'tree-changes';)/u.exec(content)?.groups ?? {};
        const { exp, named } =
          /(?<exp>export (?<named>.+) from 'tree-changes';)/u.exec(content)?.groups ?? {};
        const { exportDefault } =
          /(?<exportDefault>export \{ useTreeChanges as default \};)/u.exec(content)?.groups ?? {};

        if (imp) {
          const statement = `declare namespace UseTreeChanges {
  export ${named.replace('default as treeChanges', 'treeChanges')};

  ${exportDefault}
}

export = UseTreeChanges;`;

          return (
            content
              .replace(imp, `import ${named} from 'tree-changes';`)
              .replace(`\n${exp}`, '')
              .replace(`\n${exportDefault}`, '') + statement
          );
        }

        return false;
      },
    });

    expect(result).toMatchSnapshot();
    expect(consoleLog).toHaveBeenCalledWith('[fix-cjs-dts] âœ“ tests/.tmp/replaceContent.d.ts');
  });

  it('should return an empty array if no files are found', async () => {
    const result = await replaceContent({
      directory: 'tests/.tmp',
      name: 'fix-cjs-dts',
      pattern: '**/*.json',
      callback: () => false,
    });

    expect(consoleLog).toHaveBeenCalledWith('[fix-cjs-dts] No files matched');
    expect(result).toEqual([]);
  });

  it('should skip if the callback is falsy', async () => {
    const result = await replaceContent({
      directory: 'tests/.tmp',
      name: 'fix-cjs-dts',
      pattern: '**/*.d.ts',
      callback: () => false,
    });

    expect(result).toEqual([]);
    expect(consoleLog).toHaveBeenCalledWith('[fix-cjs-dts] skip tests/.tmp/replaceContent.d.ts');
  });
});
