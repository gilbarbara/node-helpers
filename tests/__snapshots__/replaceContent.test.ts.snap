// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replaceContent > should replace properly 1`] = `
[
  "/* eslint-disable no-restricted-exports,import/no-unresolved */
import { Data, KeyType, TreeChanges, Value, default as treeChanges } from 'tree-changes';


declare function useTreeChanges<T extends Data>(value: T): TreeChanges<KeyType<T, T>>;

declare namespace UseTreeChanges {
  export { Data, KeyType, TreeChanges, Value, treeChanges };

  export { useTreeChanges as default };
}

export = UseTreeChanges;",
]
`;
